### ISSUE TITLE
Token dispenser auth error, probably too many connections

### ISSUE THREAD

**Issue by jr3074** [matlink/gplaycli#80]
the token dispenser doesn't work now,

**Comment by matlink** [comment]
Yes, token-dispenser has auth error whatever the account it uses

**Comment by rehmatworks** [comment]
@matlink You should encourage the users to set up their own token dispenser servers, otherwise your email account gets under heavy load I feel. I've set up a dispenser server of my own and it is working pretty fine. Just checked.

**Comment by jr3074** [comment]
well, how token dispenser server works? does it try in a few time to re-login when error?
i see some news that old android version (android 2.1 or below ) cannot login to google service,
i think the real problem is new login method from Google

**Comment by matlink** [comment]
@rehmatworks well it's quite astonishing because with few other google accounts my token dispenser throws auth error.

**Comment by rehmatworks** [comment]
@jr3074 We will have to test on Android 2.x as I have been using this for Android 4.x and up. Regarding the max tries, @matlink will hopefully explain.

@matlink Do you use app-specific passwords or the regular password of the account? For me, app password is working fine. Going to test with some more email accounts right now.

**Comment by jr3074** [comment]
have you tried this?
https://github.com/dweinstein/node-google-play-cli

**Comment by matlink** [comment]
@rehmatworks No, I use the regular password. However, I created a new account, and it's working fine now on the token-dispenser. You just have to use the same URL as before, I made a little hack to bypass it.

**Comment by rehmatworks** [comment]
@jr3074 No, using gplaycli from beginning ;) @matlink Thanks for the confirmation

**Comment by jr3074** [comment]
oh my, i have tried it, and also get the same issue,
let's dig into the real issue (badauth)

**Comment by jr3074** [comment]
@rehmatworks "app password is working fine." what 'app password' is?
can you explain? thanks

**Comment by matlink** [comment]
Google have blocked these accounts I guess

**Comment by NoMore201** [comment]
I was wondering, is it possible to implement the token-dispenser into gplaycli? I was checking the token-dispenser code, and all it does (if i'm not wrong) is calling the `generateAC2DMToken` method instead of the classic `generateToken` which is implemented in gplaycli.

The only difference is that `generateAC2DMToken` requires password encryption, which is done (and explained) [here](https://github.com/yeriomin/play-store-api/blob/master/src/main/java/com/github/yeriomin/playstoreapi/PasswordEncrypter.java).
In your opinion, it is feasible to implement this?

**Comment by rehmatworks** [comment]
@jr3074 App specific password feature is enabled when you activate 2-step authentication in your Google account. You can generate app-specific passwords at https://security.google.com/settings/security/apppasswords
More information about app passwords can be found at https://support.google.com/accounts/answer/185833?hl=en

**Comment by matlink** [comment]
@NoMore201 I don't get why it would be useful. Tokendispenser allows people not to have to set a google account. If you deport the token dispenser into gplaycli, the user will have to set a google account (Which is already the case with credentials.).

**Comment by archon810** [comment]
@matlink Because then if we wanted to use our own accounts, we wouldn't need to set up a separate server, running at all times and opening up security concerns.

**Comment by matlink** [comment]
@archon810 well gplaycli already allows you to setup your own account without running tokendispenser. Check README.md.

**Comment by jr3074** [comment]
hello @matlink , the token dispenser not responding? i just test it today and got very long time response or not respon

**Comment by jr3074** [comment]
requests.exceptions.InvalidHeader: Invalid return character or leading space in header: Authorization

**Comment by NoMore201** [comment]
@jr3074 as stated before, it is better to host your own token-dispenser, to avoid problems and to not bother @matlink everytime the server goes down. I wrote simple Dockerfiles for the dispenser, If they may be helpful to you:
https://github.com/NoMore201/docker-token-dispenser

**Comment by jr3074** [comment]
@NoMore201  ya thanks,
@matlink  does badauth problem has been fixed in the latest token dispenser?
i see your discussion on https://github.com/yeriomin/token-dispenser/commit/5b9779a11a0d5e7099bae879128727fe1279b11f
he fixed it,

**Comment by matlink** [comment]
Yes it has been fixed. However, the problem persists. 
-- 
Matlink

**Comment by PedroGalindo** [comment]
Hi @matlink ,

I need to implement a private token dispenser like your system: https://matlink.fr/token/email/gplaycliacc@gmail.com

Could you help me about how implement this?

Best regards.

**Comment by NoMore201** [comment]
@PedroGalindo 
You can use @matlink fork [1] of token-dispenser and follow instruction in the README.md. Or you can check my docker-images [2], also as a reference for the installation.
[1] https://github.com/matlink/token-dispenser
[2] https://github.com/NoMore201/docker-token-dispenser

**Comment by TheT3ngu** [comment]
`Traceback (most recent call last):
  File "/usr/bin/gplaycli", line 496, in <module>
    main()
  File "/usr/bin/gplaycli", line 463, in main
    cli = GPlaycli(args, args.config)
  File "/usr/bin/gplaycli", line 92, in __init__
    self.retries = int(self.config["retries"])
KeyError: 'retries'`

Got this error whenever I trie to interact with gplaycli (update APKs, search, etc) - is this caused by the token dispenser error?

**Comment by PedroGalindo** [comment]
Hi @NoMore201 ,

Many thanks for your answer.

I'm testing with your docker (please note that I never used dockers) and I can't get the token-dispenser. I followed the next steps:
1- Clone your docker project [2]
2- I choose the plaintext option
3- I introduce my credentials in entrypoint.sh
4- Finally I try to execut this with the following comands:
 sudo docker build -t tokendispenser .
 sudo docker run -d --name tokendispenser -p 8080:8080 tokendispenser 
I need to execute this in any SO (Windows,Linux,Mac..). In my case, I try it in a Mac computer and I can't get this.

Could you help me?

Regards,
Pedro

**Comment by NoMore201** [comment]
@PedroGalindo 
First of all make sure your google account is activated with an android phone. If you created a new account but never used with android, it will not work.

Then, after running the image with docker, it will build the java code and then run it, so you need to wait some seconds for the server to be available (check the process with *htop* to see if it has finished building,  or check the logs with `sudo docker logs tokendispenser`.

If it's still not working, check out the logs with `sudo docker logs tokendispenser` to see if there's any error, and let me know :)
While the image is running, the endpoint to contact for retrieving token is http://<IP_ADDRESS>:8080/token/email/<YOUR_GOOGLE_MAIL>

**Comment by matlink** [comment]
@PedroGalindo the best way to deploy a token dispenser is to get the new token dispenser version from @yeriomin repo https://github.com/yeriomin/token-dispenser
He made a quick tutorial to make it run. Better use his version than mine that is obsolete.

**Comment by PedroGalindo** [comment]
Hi @matlink and @NoMore201 ,

Many thanks for your help.
Now it's running with @NoMore201 docker! 
Many thanks.

Best regards!

**Comment by NoMore201** [comment]
@PedroGalindo Your're welcome :)
@matlink For the images I'm using the upstream repository. And yes, the mongodb backend is quite annoying, also because if you use a recent version of mongodb it has [some problems](https://github.com/yeriomin/token-dispenser/issues/5). The solution is in that link, and also in the README.md of the docker image.

**Comment by rehmatworks** [comment]
I encourage you to set up your own Token Dispenser server. This way, @matlink's server will not get overloaded. If anyone is facing technical issues in installing and configuring Token Dispenser server, I've written a simple bash gist at https://gist.github.com/rehmatworks/9248305f0785c54672bab165a077e169 that should automate the installation of the server. Remember that port 85 will be opened on your server and plaintext file will be used to store your email and password. I encourage you to use an application specific password if possible.

**Comment by archon810** [comment]
What I don't quite understand is why the token dispensing can't just be built into gplacli without the need to run a server, which has the potential of someone else getting access to your credentials if they discover it and it's not properly firewalled?

**Comment by rehmatworks** [comment]
@archon810 Maybe there is some good amount of work in order to build this as an inbuilt feature. I don't feel that someone will be able to access your credentials as the passwords.txt file isn't accessible over HTTP.

**Comment by archon810** [comment]
That depends on where someone puts it. If they happen to stuff it in their web server somewhere, it would be.

But even if not, these tokens still give someone access to the respective Google account, no? 

I find everything related to the token server very confusing - there's no enough documentation about *how* auth works and what gplaycli does with and without the token setup.

**Comment by matlink** [comment]
No, it's only for user convenience : they have nothing to do to make gplaycli working. git clone and gplaycli works. If you want to put tokendispenser into gplaycli, it forces users to create a google account and set it up in the credentials.conf (which is already possible with the non-token-way authentication by putting credentials in this config file).

**Comment by NoMore201** [comment]
True, but in my experience token-dispenser is working very well out-of-the-box, while credentials with gplaycli not so much. Moreover, one should store credentials in a plaintext file even if using directly gplaycli without dispenser, so there's not much difference. And also if one has the token, he can authenticate as you, so password is not needed.

**Comment by rehmatworks** [comment]
@archon810 You should be taking the basic care of security issues like this. The author never expects you to put the passwords.txt file in a publicly accessible folder. Moreover, even if you need to keep the file in a place like this, still you should adjust its permissions to prevent its public access. Regarding the token, it might allow someone to make a purchase of a paid app in Google Play but that seems very unlikely. Apart from this, the token alone should not give someone access to any Google account. Still, you should use a specific gmail account for this purpose, don't link any credit card there to be safe etc.

**Comment by archon810** [comment]
@rehmatworks @matlink Is there a diagram, wiki, explanation, anything that explains how auth works and how each method differs? Specifically, what happens on a technical level with the "token" method vs the "password" method? When Google's servers are pinged in this case and why I'm getting BadAuthentication so much these days (I'm not using the token method)?

Just a really detailed primer on this would go a long way IMO.

**Comment by rehmatworks** [comment]
@archon810 What I understand about the login mechanism is that when you use the email and password combination, each time you are logged into your account whenever you make an APK download request and Google blocks your login requests when they are made quickly. With the token option, first, the token dispenser server makes an API call to Google servers and retrieves a short-lived token for you. This token is used by gplaycli to authenticate you and as of now, Google doesn't block your requests when you use the token way. @matlink may further clarify this but I feel that you should use the token way as the community (including me) is already using this without any issues and if something goes wrong with this, obviously someone will report here.

**Comment by yeriomin** [comment]
@archon810 @rehmatworks 

>Specifically, what happens on a technical level with the "token" method vs the "password" method?

Every request to Google Play servers needs to have the `Authorization` http header with a token. To get this token you have to make a request to https://android.clients.google.com/auth with Email and Passwd headers. It has worked like this for five years (probably more). There has never been any other way. Check [this](https://github.com/matlink/gplaycli/blob/master/ext_libs/googleplay_api/googleplay.py) file. It is doing the things I'm talking about in gplaycli.

A couple of years ago a new way to get the token was introduced. Almost the same, but the password needs to be hashed in a special way and put into EncryptedPasswd header instead of just putting it in plain text. No one would care if it wasn't for the fact that a month ago the old way started to **randomly** answer `Bad auth` to valid token requests. I've checked different accounts and ips from different countries - no difference. Making the same token request may or may not return a token.

The EncryptedPasswd method works properly for now.

> retrieves a short-lived token for you

It is quite long-lived. Its remains valid for months! Not sure for how long exactly.

>What I don't quite understand is why the token dispensing can't just be built into gplacli without the need to run a server

gplaycli can be modified to get the tokens using the EncryptedPasswd method. The encryption algo is [here](https://github.com/yeriomin/play-store-api/blob/master/src/main/java/com/github/yeriomin/playstoreapi/PasswordEncrypter.java). Using token dispenser lets the author keep the password to his google account secret. Otherwise it has to be stored in plaintext [here](https://github.com/matlink/gplaycli/blob/master/credentials.conf).

>But even if not, these tokens still give someone access to the respective Google account, no?

It is good for Play Store requests only, as far as I know.

**Comment by rehmatworks** [comment]
@yeriomin Thanks for this explanation. It means that we can reuse the tokens? If the answer is yes, then it would be a good practice to use the same token for APK download requests and request a new token via token dispenser server only if the old token is invalidated. This way, the hits to Google servers can be reduced.

**Comment by yeriomin** [comment]
@rehmatworks 

> It means that we can reuse the tokens?

Definitely.

> This way, the hits to Google servers can be reduced.

For the record: For the past seven days there has been 26000 requests (don't know unique or not - I'm not recording ips) to my token dispenser instance. The account I use to get tokens gets routinely blocked by google once per week. A warning comes to my email saying "Someone has your password" and the location matches the location of the data center where my token dispenser runs. Since that moment token requests are answered with `Bad auth`. After I change the password to the account it starts working properly.

**Comment by archon810** [comment]
@yeriomin @matlink Thanks for the really clear explanation. This stuff should be in docs in both token server and gplaycli, it'd help a lot of people who are confused. Think about a clean slate, a user who has no inside baseball knowledge of how it all works - they come in and see https://github.com/yeriomin/token-dispenser/blob/feb80861b6dea389cc3ba92f081148cf9bdf86b0/README.md or https://github.com/matlink/gplaycli/blob/757b6414bf5c06be15da27d729189fd32ff009ee/README.md, and they will likely be confused.

@matlink On the matter of reusing the token - does gplaycli do that? It'd make sense if it did - reuse it until it's known to be invalid, then renew. If it doesn't do that yet, it'd save on requests and time.

**Comment by jr3074** [comment]
I agree with "reusing the token", so token request only when it's invalid.

**Comment by matlink** [comment]
It does not do that yet. Please open a separate issue.

**Comment by rehmatworks** [comment]
@matlink Please consider reusing the tokens unless they become invalid. I'll create a separate issue now as several other users will be needing this feature as well.

**Comment by archon810** [comment]
Thanks for opening #81.

**Comment by matlink** [comment]
@yeriomin it means that you're changing your google account's password quite often not to get banned?

**Comment by yeriomin** [comment]
@matlink Technically the account does not get banned, but google considers it compromised and no longer gives out tokens. Web login works though, so when this happens I change the password. It happens weekly lately.
